planner/core/rule_decorrelate.go

```diff
@@ -171,10 +171,26 @@ func (s *decorrelateSolver) optimize(ctx context.Context, p LogicalPlan, opt *lo
 				// TODO: Actually, it can be optimized. We need to first push the projection down to the selection. And then the APPLY can be decorrelated.
 				goto NoOptimize
 			}
+
+			// step1: substitute the all the schema with new expressions (including correlated column maybe, but it doesn't affect the collation infer inside)
+			// eg: projection: constant("guo") --> column8, once upper layer substitution failed here, the lower layer behind
+			// projection can't supply column8 anymore.
+			//
+			//	upper OP (depend on column8)   --> projection(constant "guo" --> column8)  --> lower layer OP
+			//	          |                                                       ^
+			//	          +-------------------------------------------------------+
+			//
+			//	upper OP (depend on column8)   --> lower layer OP
+			//	          |                             ^
+			//	          +-----------------------------+      // Fail: lower layer can't supply column8 anymore.
+			hasFail := apply.columnSubstituteAll(proj.Schema(), proj.Exprs)
+			if hasFail {
+				goto NoOptimize
+			}
+			// step2: when it can be substituted all, we then just do the de-correlation.
 			for i, expr := range proj.Exprs {
 				proj.Exprs[i] = expr.Decorrelate(outerPlan.Schema())
```

Comment:
It seems like we also need to decorrelate the conditions in the `Apply` here. I guess currently failed cases are caused by this.